### PR TITLE
[WIP] Fix for snapshot action in TDataFrame

### PR DESCRIPTION
**Pull request by amadio**
**Please do not merge yet.**

This change still does not fix the problem for me, although it does fix part of the problem.

### CODE REVIEW

File: tree/treeplayer/inc/ROOT/TInterface.hxx [hunk 1]
```diff
@@ -1083,11 +1083,11 @@ protected:
...
          ForeachSlot(fillTree, {bnames[S]...});
 
-         for (auto &&tree : trees) tree->Write();
-         for (auto &&file : files) file->Write();
-
-         trees.clear();
-         files.clear();
+         for (unsigned int slot = 0; slot < nSlots; ++slot) {
+            files[slot]->cd();
+            trees[slot]->Write();
```
Comment (pcanal) [inline tree/treeplayer/inc/ROOT/TInterface.hxx]: The cd+tree->Write are redundant.  The file Write should be enough (it goes though the list of things owned by the TFile and write them).

Comment (amadio) [inline tree/treeplayer/inc/ROOT/TInterface.hxx]: Yes, Danilo already let me know. I will have only tree->Write here later.

**Comment by dpiparo** [comment]
Hi,

this mitigates the issue but still does not solve entirely the problem. Did you have the possibility to try after the fix of @pcanal for creating a TChain with a null gDirectory?
One improvement which in general we could put in place and is likely to fix the issue is to use a TDirectory::TContext to restore the status of the gDirectory to the one at the beginning of the method call. In general I would expect that the call to Snapshot does not have an influence to the state of ROOT.

**Comment by amadio** [comment]
@danilo I am still seeing some crashes and investigating the problem.

**Comment by dpiparo** [comment]
Hi @amadio , great. Did you have the possibility to try the TDirectory::TContext RAII? Perhaps this helps.

**Comment by amadio** [comment]
Yes, but it still crashes. Now in `TFile::Close()` in ~TBufferMerger() which calls `gROOT->GetListOfFiles()->Remove(this);` for `TBufferMerger::fFile`.

**Comment by amadio** [comment]
I think I could use the help of someone more familiar with what happens in the background when a file is opened, closed, etc.

**Comment by pcanal** [comment]
Do you still see the problem after removing the cd and the tree->Write()?
If so, can run with 'valgrind --suppressions=$ROOTSYS/etc/valgrind-root.supp' this should point us in the right direction.

**Comment by amadio** [comment]
@pcanal Yes, the problem is there even in the test `io/io/test/testTBufferMerger`. It's not related to TDF, but it's in the code for TBufferMerger itself. The ownership of the TTrees is unclear to me, and somehow they are destroyed more than once (or their files). I'm going to close this pull request and continue to investigate the problem. I will open a new pull request when I'm convinced the problem is fixed.

**Pull request closed without merging**